src/Spice86.Core/Emulator/Devices/ExternalInput/DualPic.cs

```diff
@@ -1,207 +1,900 @@
+        }
+
+        if (pic.CurrentIcwIndex++ >= pic.RemainingInitializationWords) {
+            pic.CurrentIcwIndex = 0;
+        }
+    }
+
+    /// <summary>
+    ///     Asserts a numbered IRQ and updates controller state accordingly.
+    /// </summary>
+    /// <param name="irq">IRQ number in the range 0–15.</param>
+    public void ActivateIrq(byte irq) {
+        if (irq > MaxIrq) {
+            _logger.Error("PIC: Activation requested for out-of-range IRQ {Irq}", irq);
+            return;
+        }
+
+        byte controllerIrq = GetEffectiveControllerIrq(irq);
+
+        int oldCycles = _cpuState.Cycles;
+        GetPicByIrq(irq).RaiseIrq(controllerIrq);
+
+        if (oldCycles == _cpuState.Cycles) {
+            return;
+        }
+
+        // if CPU_Cycles have changed, this means that the interrupt was triggered by an I/O
+        // register writing rather than an event.
+        // Real hardware executes 0 to ~13 NOPs or comparable instructions
+        // before the processor picks up the interrupt. Let's try with 2
+        // cycles here.
+        // Required by Panic demo (irq0), It came from the desert (MPU401)
+        // Does it matter if CPU_CycleLeft becomes negative?
+        // It might be an idea to do this always to simulate this
+        // So on writing mask and EOI as well. (so inside the activate function)
+        //		CPU_CycleLeft += (CPU_Cycles-2);
+        _cpuState.CyclesLeft -= 2;
+        _cpuState.Cycles = 2;
+    }
+
+    /// <summary>
+    ///     Deasserts a numbered IRQ.
+    /// </summary>
+    /// <param name="irq">IRQ number in the range 0–15.</param>
+    public void DeactivateIrq(byte irq) {
+        if (irq > MaxIrq) {
+            _logger.Error("PIC: Deactivation requested for out-of-range IRQ {Irq}", irq);
+            return;
+        }
+
+        byte controllerIrq = GetEffectiveControllerIrq(irq);
+        GetPicByIrq(irq).LowerIrq(controllerIrq);
+    }
+
+    /// <summary>
+    ///     Maps a global IRQ number to the controller-relative line serviced by the PIC.
+    /// </summary>
+    /// <param name="irq">IRQ number in the range 0–15.</param>
+    /// <returns>The IRQ index relative to the owning controller.</returns>
+    private static byte GetEffectiveControllerIrq(byte irq) {
+        return irq > 7 ? (byte)(irq - 8) : irq;
+    }
+
+    // Select the first pending IRQ on the secondary controller and cascade it through IRQ 2.
+    private void SecondaryStartIrq() {
+        byte selectedIrq = _secondaryPic.GetNextPendingIrq();
+        if (selectedIrq == NoPendingIrq) {
+            _logger.Error("PIC {Controller}: IRQ 2 is active, but IRQ is not active on the secondary controller.",
+                nameof(PicController.Secondary));
+
+            return;
+        }
+
+        _secondaryPic.StartIrq(selectedIrq);
+        _primaryPic.StartIrq(2);
+        _cpuState.CpuHwInterrupt((byte)(_secondaryPic.InterruptVectorBase + selectedIrq));
```

Comment:
Why setting this to a separate state object since this is going to be returned by the caller (ComputeVectorNumber)
This is also called from RunQueue but this should not do anything more than running the event queue and setup the PIC if an event for that is there